builder/dockerfile/parser/parser.go

```diff
@@ -32,31 +33,40 @@ type Node struct {
 	Attributes map[string]bool // special attributes for this node
 	Original   string          // original line used before parsing
 	Flags      []string        // only top Node should have this set
-	StartLine  int             // the line in the original dockerfile where the node begins
-	EndLine    int             // the line in the original dockerfile where the node ends
+	startLine  int             // the line in the original dockerfile where the node begins (used in tests)
+	endLine    int             // the line in the original dockerfile where the node ends (used in tests)
 }
 
-// Directive is the structure used during a build run to hold the state of
+// Directives is the structure used during a build run to hold the state of
 // parsing directives.
-type Directive struct {
-	EscapeToken           rune           // Current escape token
-	LineContinuationRegex *regexp.Regexp // Current line contination regex
-	LookingForDirectives  bool           // Whether we are currently looking for directives
-	EscapeSeen            bool           // Whether the escape directive has been seen
+type Directives struct {
+	EscapeToken           rune                // Current escape token
+	LineContinuationRegex *regexp.Regexp      // Current line contination regex
+	usedDirectives        map[string]struct{} // Whether the escape directive has been seen
 }
 
 var (
-	dispatch           map[string]func(string, *Directive) (*Node, map[string]bool, error)
-	tokenWhitespace    = regexp.MustCompile(`[\t\v\f\r ]+`)
-	tokenEscapeCommand = regexp.MustCompile(`^#[ \t]*escape[ \t]*=[ \t]*(?P<escapechar>.).*$`)
-	tokenComment       = regexp.MustCompile(`^#.*$`)
+	dispatch        map[string]func(string, Directives) (*Node, map[string]bool, error)
+	tokenWhitespace = regexp.MustCompile(`[\t\v\f\r ]+`)
+	directiveRegexp = regexp.MustCompile(`^#[ \t]*([a-z0-9]+)[ \t]*=[ \t]*(.*)$`)
+	tokenComment    = regexp.MustCompile(`^#.*$`)
+	errNoDirective  = errors.New("not a directive")
 )
 
 // DefaultEscapeToken is the default escape token
 const DefaultEscapeToken = "\\"
 
+// DefaultDirectives returns directives struct with default properties
+func DefaultDirectives() Directives {
+	d := &Directives{usedDirectives: make(map[string]struct{})}
+	if err := d.SetEscapeToken(DefaultEscapeToken); err != nil {
+		panic(err)
+	}
+	return *d
+}
+
 // SetEscapeToken sets the default token for escaping characters in a Dockerfile.
-func SetEscapeToken(s string, d *Directive) error {
+func (d *Directives) SetEscapeToken(s string) error {
 	if s != "`" && s != "\\" {
 		return fmt.Errorf("invalid ESCAPE '%s'. Must be ` or \\", s)
```

Comment:
I thought this was swallowed (by design) before and just ignored, the whole line being treated as a comment.